### PR TITLE
Add ability to double click to select all connected components

### LINKED ISSUE #800: Double click on a component to select all connected components
Similar to how double clicking on a wire will select all of the wire's components and nodes splitting it, it could be useful to be able to double click a component and select all other components connected to it.

This would essentially select all the components necessary to create an IC containing the component that was double clicked, which could be easier than selecting those components individually (particularly if the canvas contains many other, non-connected components in the way, making bounding box highlighting more tedious. This also would seem to be the easiest method for users on a trackpad, without a mouse).

### PR DESCRIPTION
**Pull request by mercuryhg31**
Closes #800

### CODE REVIEW

**Review by TGCrystal — APPROVED** [review]

File: src/app/core/utils/ComponentUtils.ts [hunk 1]
```diff
@@ -136,6 +136,38 @@ export function GetPath(w: Wire | Node): Array<Wire | Node> {
...
+    const queue = new Array<Component>(c);
+    const visited = new Set<Component>();
+
+    while(queue.length > 0) {
+        const q = queue.shift()!;
+
+        visited.add(q);
+        path.push(q);
+        for (let p of q.getPorts()) {
+            for (let w of p.getWires()) {
```
Comment (LeonMontealegre) [inline src/app/core/utils/ComponentUtils.ts:159]: nit: use `const`

File: src/app/core/tools/handlers/SelectPathHandler.ts [hunk 2]
```diff
@@ -27,8 +34,17 @@ export const SelectPathHandler: EventHandler = ({
...
+            .find(o => o.isWithinSelectBounds(worldMousePos));
 
-        history.add(CreateGroupSelectAction(selections, path).execute());
+        if (wire != null) {
+            const wirePath = GetWirePath(wire!);
+            history.add(CreateGroupSelectAction(selections, wirePath).execute());
+        } else {
+            const objPath = GetComponentPath(component!);
+            history.add(CreateGroupSelectAction(selections, objPath).execute())
+        }
```
Comment (LeonMontealegre) [inline src/app/core/tools/handlers/SelectPathHandler.ts:48]: Can condense this as
```ts
const path = (wire) ? (GetWirePath(wire)) : (GetComponentPath(component!));
history.add(CreateGroupSelectAction(selections, path).execute());
```

File: src/app/core/tools/handlers/SelectPathHandler.ts [hunk 3]
```diff
@@ -27,8 +34,17 @@ export const SelectPathHandler: EventHandler = ({
             .getWires()
             .reverse()
             .find(o => o.isWithinSelectBounds(worldMousePos));
-        const path = GetPath(wire!);
+        const component = designer
+            .getObjects()
+            .reverse()
+            .find(o => o.isWithinSelectBounds(worldMousePos));
```
Comment (LeonMontealegre) [inline src/app/core/tools/handlers/SelectPathHandler.ts:40]: Combine these using `getAll()`, which, when reversed, will have the wires in front and the objects at the back which is the desired behavior. (also make sure to cast `as (Component | Wire)`)

Then for the if-statement below, use `(obj instanceof Wire)`

**Review by LeonMontealegre — CHANGES_REQUESTED** [review]

File: src/app/core/tools/handlers/SelectPathHandler.ts [hunk 4]
```diff
@@ -2,33 +2,36 @@ import {LEFT_MOUSE_BUTTON} from "core/utils/Constants";
...
+               o.isWithinSelectBounds(camera.getWorldPos(input.getMousePos())))
+           !== undefined
+        ),
 
     getResponse: ({input, camera, history, designer, selections}: CircuitInfo) => {
         const worldMousePos = camera.getWorldPos(input.getMousePos());
 
-        const wire = designer
-            .getWires()
+        const obj: Component | Wire = designer
```
Comment (LeonMontealegre) [inline src/app/core/tools/handlers/SelectPathHandler.ts:28]: nit: Shouldn't need a double cast, the cast below should be enough

File: src/app/core/tools/handlers/SelectPathHandler.ts [hunk 5]
```diff
@@ -2,33 +2,36 @@ import {LEFT_MOUSE_BUTTON} from "core/utils/Constants";
...
 import {CircuitInfo} from "core/utils/CircuitInfo";
-import {GetPath} from "core/utils/ComponentUtils";
+import {GetComponentPath, GetWirePath} from "core/utils/ComponentUtils";
 
 import {CreateGroupSelectAction} from "core/actions/selection/SelectAction";
 
 import {EventHandler} from "../EventHandler";
 
+import {Component} from "core/models";
+import {Wire} from "core/models";
```
Comment (LeonMontealegre) [inline src/app/core/tools/handlers/SelectPathHandler.ts:12]: nit: Combine these to `import {Component, Wire} from "core/models";` and more the import to be above the ".." import

**Review by LeonMontealegre — CHANGES_REQUESTED** [review]
Just a few nits!

**Review by TGCrystal — APPROVED** [review]

**Review by LeonMontealegre — APPROVED** [review]

**Pull request merged**